rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,485 @@
+             updating, and what changes to apply.
+
+* **Provider Registry**: An index of available Terraform _providers_.
+
+* **Terraform Cloud**: A hosted service run by Hashicorp that
+  * hosts public and private _Provider Registries_
+  * stores Terraform _State_, and
+  * remotely executes Terraform scripts
+
+* **Terraform Enterprise**: A privately-hosted version of _Terraform Cloud_.
+
+* **Public Registry**: A publicly-accessible _registry_ that does not require
+                       authentication to access.
+
+* **Private Registry**: A _registry_ that requires authentication to access.
+
+* **Custom Registry**: Any _registry_ not hosted by _Teleport Cloud_
+                       or _Teleport Enterprise_. A custom implementation of
+                       the Terraform Registry Protocol (see below).
+
+The target of this RFD is, by the above definitions, a public, custom provider
+registry.
+
+### Terraform provider naming
+
+Terraform providers have a two-part naming convention of `namespace/provider`,
+where _namespace_ is the publishing organization and _provider_  is the the
+actual name of the provider itself.
+
+In this document I have used `gravitational` as the namespace and `teleport` as
+the provider in order to disambiguate the two components of the provider name,
+but there is no reason that we can't use `teleport/teleport` in the actual
+deployment if that's the preferred branding.
+
+## References
+
+* [Cloud RFD 0004 - Release Asset Management](https://github.com/gravitational/cloud/blob/master/rfd/0004-Release-Asset-Management.md)
+* [Cloud RFD 0017 - Artifact Storage Standards](https://github.com/gravitational/cloud/blob/master/rfd/0017-artifact-storage-standards.md)
+* [Terraform Provider Registry Specification](https://www.terraform.io/internals/provider-registry-protocol)
+
+## Details
+
+### Terraform Registry Protocol requirements
+
+For the full specification, see the [Provider Registry Specification](https://www.terraform.io/internals/provider-registry-protocol).
+
+The major points of the registry protocol affecting this document are
+
+1. A web service that exposes the registry index,
+2. The binary packages referenced by the registry must be packaged in zipfiles,
+   as opposed to the compressed tarballs our current build process produces, and
+3. The checksum files must be signed with a GPG key
+
+### Proposed Implementation
+
+Although the Provider Registry Specification is written in terms of a web
+service that will dynamically generate the appropriate responses, it is
+_entirely possible_ to generate a registry index as a simple directory tree and
+then statically serve those files to a client. This approach dramatically
+simplifies the process of applying updates to the registry, and is therefore
+the approach we will consider here.
+
+#### Why can't we just use Houston?
+
+The existing Houston distribution tool at [get.gravitational.com](https://get.gravitational.com/)
```

Comment:
@tcsc @pierrebeaucamp Houston does not support that, but the design below looks very similar to our CDN project, any reason we can't serve this asset from our CDN and support additional URL format to support terraform requirements?

Reply:
@tcsc I also did not realize that you've mostly implemented this RFD here:

https://github.com/gravitational/teleport-plugins/pull/457/files

I would not block us on consolidating it right now, but may be merge this code at some point in the CDN in the future version for unification?